In [1]:
import pandas as pd
import numpy as np

1. Creacion de Dataframes

In [ ]:
df1 = pd.DataFrame({"numeros": np.arange(20)})
df1
df2 = pd.DataFrame({"aleatorios": np.random.rand(10)})
df2
df3 = df3 = pd.concat([df1, df2], axis=1)
df3
df3.to_csv("df3.csv")

2. Indexacion y seleccion 

In [25]:
df = pd.DataFrame({
    "A": range (1, 11),
    "B": [5,3,6,8,2,9,1,7,4,10]
})
df
df.loc[df["B"] > 5]
df.loc[df["B"] < 4, "B"] = pd.NA
df

A     B
0   1   5.0
1   2   NaN
2   3   6.0
3   4   8.0
4   5   NaN
5   6   9.0
6   7   NaN
7   8   7.0
8   9   4.0
9  10  10.0

In [ ]:
3. Analisis de Datos

In [20]:
df = pd.read_csv("winequality-red.csv", delimiter = ";")
df
df.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1599.000000       1599.000000  1599.000000     1599.000000   
mean        8.319637          0.527821     0.270976        2.538806   
std         1.741096          0.179060     0.194801        1.409928   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1599.000000          1599.000000           1599.000000  1599.000000   
mean      0.087467            15.874922             46.467792     0.996747   
std       0.047065            10.460157             32.895324     0.001887   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             22.000000     0.995600   
50%       0.079000            14.000000             38.000000     0.996750   
75%       0.090000            21.000000             62.000000     0.997835   
max       0.611000            72.000000            289.000000     1.003690   

                pH    sulphates      alcohol      quality  
count  1599.000000  1599.000000  1599.000000  1599.000000  
mean      3.311113     0.658149    10.422983     5.636023  
std       0.154386     0.169507     1.065668     0.807569  
min       2.740000     0.330000     8.400000     3.000000  
25%       3.210000     0.550000     9.500000     5.000000  
50%       3.310000     0.620000    10.200000     6.000000  
75%       3.400000     0.730000    11.100000     6.000000  
max       4.010000     2.000000    14.900000     8.000000

Fixed Acidity:
Los vinos tintos tienen un nivel de 8,32 fixed acidity en promedio, el mínimo acidity es 4,6 el máximo es 15.9.
75% de los vinos tienen un nivel de fixed acidity inferior a 9,2. Desviacion estandar es bajo, 1,74 uniades en general del valor medio (8,32)

Volatile Acidity:
Los valores varían entre 0.12 y 1.58, con poca variabilidad (desviación estándar de 0.17).

Citric Acid:
El promedio de ácido cítrico es 0,27. Existen vinos con niveles de ácido cítrico igual a 0.
El 75% de los vinos tiene un nivel de ácido cítrico igual o inferior a 0,42, lo que indica que los vinos con valores cercanos a 1 son poco frecuentes.

Residual Sugar:
El promedio es 2,5, valor maximo es 15.5 con desviación estándar 1,4. Significa que hay vinos con nivel de residual sugar 15.5, pero son muy poco frecuentes. 75% de los vinos tienen un nivel de Residual sugar igual o inferior a 2,6.

Chlorides:
Los vinos son muy similares en el nivel de cholides. Promedio: 0.08. Valores generalmente bajos que van de 0.01 a 0.06 con una desviación estándar muy pequeña, lo que indica poca variabilidad.

Free sulfur: 
Promedio: 15.8. Std 10,46. Valores entre 1 y 72. Variabilidad alta, 75% de los vinos con valores igual o inferior a 21. Unos pocos vinos tienen nivel de Free Sulfur entre 21 y 72.

Total Sulfur dioxide:
Aquí es donde se observa la mayor variabilidad entre los vinos, con valores que van desde 6 hasta 289.
La desviación estándar es 32,89, lo que indica que, aunque el promedio es 46, muchos vinos presentan niveles de total sulfur dioxide mucho más bajos o mucho más altos.

Density:
Los valores de densidad son muy similares entre los vinos; todos se encuentran entre 0.99 y 1.00.
No hay mucha variabilidad en esta característica.

pH:
Promedio 3,3 con std 0.15. No hay mucha diferencia en el nivel de pH entre los vinos.

Sulphates:
Promedio 0.65 con std 0.17. No hay mucha diferencia en el nivel de sulphates entre los vinos, pero ejemplo en el data de vino con nivel 2.0 que presente un valor atipico.

Alcohol:
Promedio: 10.4 con std 1.06. Valor minimo 8,4, maximo 14.9. 50% de los vinos tinen un % de alcohol igual o menos quw 10.2%

Quality:
Una variable que no miden y que es más subjetiva. Los vinos con peor calidad tienen una puntuación de 3, mientras que los mejores vinos alcanzan una calidad de 8. El promedio es 5.63 con una desviación estándar de 0.8 (baja). Solo el 25% de los vinos tiene una puntuación de 7 o superior.
df.loc[df["quality"] <= 4, "quality"] = "malo"
df.loc[(df["quality"] > 4) & (df["quality"] <= 6), "quality"] = "regular"
df.loc[df["quality"] > 6, "quality"] = "bueno"

In [24]:
df["quality"] = df["quality"].astype(str)
df.loc[df["quality"].isin(["3", "4"]), "quality"] = "malo"
df.loc[df["quality"].isin(["5", "6"]), "quality"] = "regular"
df.loc[df["quality"].isin(["7", "8"]), "quality"] = "bueno"

df.tail(15)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
1584            6.7             0.320         0.44             2.4      0.061   
1585            7.2             0.390         0.44             2.6      0.066   
1586            7.5             0.310         0.41             2.4      0.065   
1587            5.8             0.610         0.11             1.8      0.066   
1588            7.2             0.660         0.33             2.5      0.068   
1589            6.6             0.725         0.20             7.8      0.073   
1590            6.3             0.550         0.15             1.8      0.077   
1591            5.4             0.740         0.09             1.7      0.089   
1592            6.3             0.510         0.13             2.3      0.076   
1593            6.8             0.620         0.08             1.9      0.068   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
1584                 24.0                  34.0  0.99484  3.29       0.80   
1585                 22.0                  48.0  0.99494  3.30       0.84   
1586                 34.0                  60.0  0.99492  3.34       0.85   
1587                 18.0                  28.0  0.99483  3.55       0.66   
1588                 34.0                 102.0  0.99414  3.27       0.78   
1589                 29.0                  79.0  0.99770  3.29       0.54   
1590                 26.0                  35.0  0.99314  3.32       0.82   
1591                 16.0                  26.0  0.99402  3.67       0.56   
1592                 29.0                  40.0  0.99574  3.42       0.75   
1593                 28.0                  38.0  0.99651  3.42       0.82   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  
1584     11.6    bueno  
1585     11.5  regular  
1586     11.4  regular  
1587     10.9  regular  
1588     12.8  regular  
1589      9.2  regular  
1590     11.6  regular  
1591     11.6  regular  
1592     11.0  regular  
1593      9.5  regular  
1594     10.5  regular  
1595     11.2  regular  
1596     11.0  regular  
1597     10.2  regular  
1598     11.0  regular

In [35]:
df = df[df["quality"] != "malo"]
df.describe
df["quality"].unique()
df["quality"].value_counts()

quality
regular    1319
bueno       217
Name: count, dtype: int64

In [37]:
df["citric acid"] = df["citric acid"] * 100 
df = df.rename(columns={"citric acid": "citric acid %"})
df

fixed acidity  volatile acidity  citric acid %  residual sugar  \
0               7.4             0.700            0.0             1.9   
1               7.8             0.880            0.0             2.6   
2               7.8             0.760          400.0             2.3   
3              11.2             0.280         5600.0             1.9   
4               7.4             0.700            0.0             1.9   
...             ...               ...            ...             ...   
1594            6.2             0.600          800.0             2.0   
1595            5.9             0.550         1000.0             2.2   
1596            6.3             0.510         1300.0             2.3   
1597            5.9             0.645         1200.0             2.0   
1598            6.0             0.310         4700.0             3.6   

      chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
0         0.076                 11.0                  34.0  0.99780  3.51   
1         0.098                 25.0                  67.0  0.99680  3.20   
2         0.092                 15.0                  54.0  0.99700  3.26   
3         0.075                 17.0                  60.0  0.99800  3.16   
4         0.076                 11.0                  34.0  0.99780  3.51   
...         ...                  ...                   ...      ...   ...   
1594      0.090                 32.0                  44.0  0.99490  3.45   
1595      0.062                 39.0                  51.0  0.99512  3.52   
1596      0.076                 29.0                  40.0  0.99574  3.42   
1597      0.075                 32.0                  44.0  0.99547  3.57   
1598      0.067                 18.0                  42.0  0.99549  3.39   

      sulphates  alcohol  quality  
0          0.56      9.4  regular  
1          0.68      9.8  regular  
2          0.65      9.8  regular  
3          0.58      9.8  regular  
4          0.56      9.4  regular  
...         ...      ...      ...  
1594       0.58     10.5  regular  
1595       0.76     11.2  regular  
1596       0.75     11.0  regular  
1597       0.71     10.2  regular  
1598       0.66     11.0  regular  

[1536 rows x 12 columns]

In [42]:
df.sort_values(by="pH", ascending=False).head(7)

fixed acidity  volatile acidity  citric acid %  residual sugar  \
1316            5.4             0.740            0.0            1.20   
1321            5.0             0.740            0.0            1.20   
695             5.1             0.470          200.0            1.30   
95              4.7             0.600         1700.0            2.30   
1111            5.4             0.420         2700.0            2.00   
1300            5.2             0.645            0.0            2.15   
553             5.0             1.040         2400.0            1.60   

      chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
1316      0.041                 16.0                  46.0  0.99258  4.01   
1321      0.041                 16.0                  46.0  0.99258  4.01   
695       0.034                 18.0                  44.0  0.99210  3.90   
95        0.058                 17.0                 106.0  0.99320  3.85   
1111      0.092                 23.0                  55.0  0.99471  3.78   
1300      0.080                 15.0                  28.0  0.99444  3.78   
553       0.050                 32.0                  96.0  0.99340  3.74   

      sulphates  alcohol  quality  
1316       0.59     12.5  regular  
1321       0.59     12.5  regular  
695        0.62     12.8  regular  
95         0.60     12.9  regular  
1111       0.64     12.3    bueno  
1300       0.61     12.5  regular  
553        0.62     11.5  regular

In [2]:
import pandas as pd
import numpy as np
import Operaciones_Practica2.py


SyntaxError: invalid syntax (Operaciones_Practica2.py, line 2)